### PR TITLE
Bring `visitorKeys` back

### PR DESCRIPTION
**Pull request by ardatan**
See https://github.com/graphql/graphql-js/issues/3245#issuecomment-907177003

I checked `Allow edits by maintainers`. So you can make changes on this PR instead of creating a new one by checking out this branch using GitHub CLI. Also you can leave comments here then I can also do the changes you want :)

### CODE REVIEW

**Comment by yaacovCR** [comment]
Returning false means you no longer visit any of the children, what if you want to visit only some?

**Comment by yaacovCR** [comment]
@IvanGoncharov  why exactly was it deprecated even in terms of its original rationale of custom ast?

in general, for use cases above, it lets you whitelist what you want to visit rather than entering each child node with a visitor function that returns false, the latter being slightly less performant, and sometimes quite more verbose.

**Comment by yaacovCR** [comment]
Related: https://github.com/ardatan/graphql-tools/pull/3481

**Comment by yaacovCR** [comment]
we can workaround the above bug secondary to lack of visitorKeys, but it doesn't change that we are doing too much visiting :(

we want to be speedy; if visitorKeys is dropped upstream, we will have to re-implement the v15 version -- please keep

File: src/language/__tests__/visitor-test.ts [hunk 1]
```diff
@@ -893,6 +894,41 @@ describe('Visitor', () => {
...
+    const visitor: ASTVisitor = {
+      enter(node) {
+        visited.push(['enter', node.kind, getValue(node)]);
+      },
+      leave(node) {
+        visited.push(['leave', node.kind, getValue(node)]);
+      },
+    };
+
+    const exampleDocumentAST = parse(/* GraphQL */ `
```
Comment (IvanGoncharov) [inline src/language/__tests__/visitor-test.ts:914]: We don't use `/* GraphQL */` tag in our source code.
All tests in this file name their variable simply `ast`.
Always try to match the style of another test.

File: src/language/__tests__/visitor-test.ts [hunk 2]
```diff
@@ -893,6 +894,41 @@ describe('Visitor', () => {
...
+      query ExampleOperation {
+        someField
+      }
+    `);
+
+    visit(exampleDocumentAST, visitor, visitorKeyMap);
+
+    expect(visited).to.deep.equal([
+      ['enter', Kind.DOCUMENT, undefined],
```
Comment (IvanGoncharov) [inline src/language/__tests__/visitor-test.ts:923]: All other tests use literal values instead of `Kind` properties.

File: src/language/__tests__/visitor-test.ts [hunk 3]
```diff
@@ -893,6 +894,41 @@ describe('Visitor', () => {
     ]);
   });
 
+  it('visits only the specified `Kind` in visitorKeyMap', () => {
+    const visited: Array<[string, string, ReturnType<typeof getValue>]> = [];
```
Comment (IvanGoncharov) [inline src/language/__tests__/visitor-test.ts:898]: All other test use `Array<any>`.
If you want to improve typings it should be done in a separate PR and for the whole file.

File: src/language/visitor.ts [hunk 4]
```diff
@@ -76,9 +76,14 @@ type ReducedField<T, R> = T extends null | undefined
   ? ReadonlyArray<R>
   : R;
 
-const QueryDocumentKeys = {
-  Name: [],
+/**
+ * A KeyMap describes each the traversable properties of each kind of node.
+ */
+export type VisitorKeyMap<KindToNode> = {
```
Comment (IvanGoncharov) [inline src/language/visitor.ts:82]: Since we dropped support for custom ASTs we don't need `KindToNode` argument also I renamed `VisitorKeyMap` to `ASTVisitorKeyMap` and will provide an alias for `v15`.

**Comment by IvanGoncharov** [comment]
> I checked Allow edits by maintainers. So you can make changes on this PR instead of creating a new one by checking out this branch using GitHub CLI. Also you can leave comments here then I can also do the changes you want :)

@ardatan That's great, however, I created new PR for a different reason.
In this and your previous PRs, you add code that doesn't match style of the whole file.
A test case that you added in this PR gives a good example of what I'm talking about.
All other test cases follow the same patterns and the same name convention so I expect one that you added to also do this.
I'm 100% up for providing necessary context when reviewing PR but I expect the PR authors to review files they are changing and code around the change.
Without that, I need to check every line of the proposed change and match it with the rest of the code (very hard in GitHub review UI) and spend time left review comments.
And that is a pretty time-consuming and exhausting process, so it's simpler for me to write code from scratch and create new PR.
For example, even though this PR is added single test case it took me way more time to review it and leave comments than to simply from it from scratch using other test cases as templates.

This situation repeats itself and was the major issue with all your previous PRs.
So if you want your future PRs to be merged please spend time and try to match the code style and conventions used in this project. If you to improve something (e.g. typings) do it in separate PR and for all places that have the pattern, you want to change.

P.S. I applied changes to this PR and merging it, but the above comment stays true for future PRs.

**Comment by IvanGoncharov** [comment]
@yaacovCR I'm merging this one but I think it's not a good long-term solution to expose such a low-level mechanism and also think it's brittle and non-intuitive API. So it should be a part of #3225 and want to continue the discussion there.

**Comment by IvanGoncharov** [comment]
@ardatan Added `ASTVisitorKeyMap` to v15 see #3267 will release new version of v15 shortly

**Comment by leszekhanusz** [comment]
Another use for the visitorKeys is present in graphql-python/gql to [parse the results coming from the backend](https://github.com/graphql-python/gql/blob/47aaf92c995bcdfbdda38ffc9997bae6d64da479/gql/utilities/parse_result.py#L37)
